processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -251,6 +245,120 @@ public static boolean convertSegment(File toConvert, File converted, IndexSpec i
+      final int[] dim1Vals = dims1[i];
+      final int[] dim2Vals = dims2[i];
+      final String dim1Name = dim1Names.get(i);
+      final String dim2Name = dim2Names.get(i);
+      final Indexed<String> dim1ValNames = adapter1.getDimValueLookup(dim1Name);
+      final Indexed<String> dim2ValNames = adapter2.getDimValueLookup(dim2Name);
+
+      if (dim1Vals == null || dim2Vals == null) {
+        if (dim1Vals != dim2Vals) {
+          throw new SegmentValidationException(
+              "Expected nulls, found %s and %s",
+              Arrays.toString(dim1Vals),
+              Arrays.toString(dim2Vals)
+          );
+        } else {
+          continue;
+        }
+      }
+      if (dim1Vals.length != dim2Vals.length) {
+        if (dim1Vals.length == 0 || dim2Vals.length == 0) {
+          // Might be OK if one of them has null. This occurs in both IndexMakerTest and in IndexIOTest
+          // At this point at least dim1Vals or dim2Vals is an empty array.
+          if (dim1Vals.length == 0 && dim2Vals.length == 1) {
+            final String dimValName = dim2ValNames.get(dim2Vals[0]);
+            if (!(dimValName == null)) {
+              throw new SegmentValidationException(
+                  "Dim [%s] value [%s] is not null",
+                  dim2Name,
+                  dimValName
+              );
+            }
+          } else if (dim2Vals.length == 0 && dim1Vals.length == 1) {
+            final String dimValName = dim1ValNames.get(dim1Vals[0]);
+            if (!(dimValName == null)) {
+              throw new SegmentValidationException(
+                  "Dim [%s] value [%s] is not null",
+                  dim1Name,
+                  dimValName
+              );
+            }
+          } else {
+            // There is probably some confusing boolean statement that could be used to make this exception and
+            // the one below appear both be used for the same thing. But I think its a bit cleaner just to put it here.
+            throw new SegmentValidationException(
+                "Dim [%s] value lengths not equal. Expected %d found %d",
+                dim1Name,
+                dims1.length,
+                dims2.length
+            );
+          }
+        } else {
+          throw new SegmentValidationException(
+              "Dim [%s] value lengths not equal. Expected %d found %d",
+              dim1Name,
+              dims1.length,
+              dims2.length
+          );
+        }
+      }
+
+      for (int j = 0; j < Math.max(dim1Vals.length, dim2Vals.length); ++j) {
+        final int dIdex1 = dim1Vals.length <= j ? -1 : dim1Vals[j];
+        final int dIdex2 = dim2Vals.length <= j ? -1 : dim2Vals[j];
+
+        if (dIdex1 == dIdex2) {
+          continue;
+        }
+
+        final String dim1ValName = dIdex1 < 0 ? null : dim1ValNames.get(dIdex1);
+        final String dim2ValName = dIdex2 < 0 ? null : dim2ValNames.get(dIdex2);
+        if ((dim1ValName == null) || (dim2ValName == null)) {
```

Comment:
I personally find it easier to read when the parentheses are explicitly placed in chained boolean operators. Less brain cycles to understand where the splits occur.